feat: UIG-2909 - vl-button-next - introductie component

ksvlaanderen opened the pull request:
vl-button-next toegevoegd
Storybook en Cypress
Form demo voorbeeld aangepast
Form data voorbeeld aangepast
Form custom validation voorbeeld aangepast
Playground-lit form voorbeeld aangepast

[Jira](https://www.milieuinfo.be/jira/browse/UIG-2909)
[Bamboo](https://www.milieuinfo.be/bamboo/browse/UIGOV-CUWC342)

Goldflow commented on libs/components/src/next/button/vl-button.css.ts:
```diff
@@ -0,0 +1,234 @@
+import { waveAnimation } from '@domg-wc/common-utilities/css/mixin/animation.mixin.css';
+import { CSSResult, css, unsafeCSS } from 'lit';
+
+const borderWidth = '0.2rem';
```
Als dit `const borderWidth = '0.2rem';` naar `const borderWidth = 0.2;` verandert, hoef je geen `unsafeCSS()` te gebruiken.

ksvlaanderen replied:
Ik heb gebruik gemaakt van de '0,2rem' string zodat ik deze makkelijk in de calc() css methode kan gebruiken.
Het lijkt me ook beter om de value en de eenheid op dezelfde plaats te definiëren.

Goldflow commented on libs/components/src/next/button/vl-button.css.ts:
```diff
@@ -0,0 +1,234 @@
+        &.secondary {
+            color: var(--vl-action-color);
+            background-color: transparent;
+            border: ${unsafeCSS(borderWidth)} solid currentColor;
```
Als bovenaan `const borderWidth = '0.2rem';` naar `const borderWidth = 0.2;` verandert, hoef je geen `unsafeCSS()` te gebruiken.

ksvlaanderen replied:
Zie reactie hierboven

Goldflow commented on libs/components/src/next/button/vl-button.css.ts:
```diff
@@ -0,0 +1,234 @@
+
+const borderWidth = '0.2rem';
+const borderWidthSmall = '0.1rem';
+const differenceBetweenRegularBorderAndSmallBorder = `calc(${borderWidth} - ${borderWidthSmall})`;
```
Gelijkaardig aan vorige opmerking, zou je hier ook die `calc(${borderWidth}rem - ${borderWidthSmall})rem)` kunnen schrijven waar nu `differenceBetweenRegularBorderAndSmallBorder` wordt gebruikt om `unsafeCSS()` te vermijden. Moet niet.

ksvlaanderen replied:
Zie reactie hierboven.

Goldflow commented on libs/components/src/next/button/vl-button.component.ts:
```diff
@@ -0,0 +1,130 @@
+                hasChanged: (_value, oldValue) => {
+                    if (oldValue === undefined) {
+                        return false;
+                    }
```
Kan je toelichten waarom `hasChanged()` `false` moet teruggeven wanneer `oldValue` `undefined` is? Vermoed dat er wel voor een goede reden staat

ksvlaanderen replied:
Done

Goldflow commented on libs/components/src/next/button/stories/vl-button.stories-arg.ts:
```diff
@@ -0,0 +1,160 @@
+    },
+    secondary: {
+        name: 'secondary',
+        description: 'Beeldt de button af als een secondary button.',
```
Eventueel toelichten wat bedoeld wordt met `secondary` en `tertiary`, meerbepaald dat het een variant in styling betreft.

bv.: `Beeldt de button af in secondary button stijl.`

ksvlaanderen replied:
Wordt uitgebreider uitgelegd op de docs pagina.

Goldflow commented on libs/components/src/next/button/vl-button.component.ts:
```diff
@@ -0,0 +1,130 @@
+
+    static get properties(): PropertyDeclarations {
+        return {
+            type: { type: String },
```
Eventueel ook `autofocus` ondersteunen (mss eens afchecken in welke mate dit door afnemers ook gebruikt wordt in opengrok).

In principe is het met een native button ook mogelijk om een `value` mee te geven. Moeten hier zeker geen FormControl van maken maar mss wel toelaten om `value` in te stellen?

ksvlaanderen replied:
Voorlopig houden zoals het is, als er een use-case van de afnemers komt voegen we het dan toe.

Goldflow commented on libs/components/src/next/button/vl-button.component.ts:
```diff
@@ -0,0 +1,130 @@
+        if (this.type === 'reset') {
+            this.closest('form')?.reset();
+        }
+
```
Op zich is het ook mogelijk op een native button `form`-attribuut mee te geven, die het mogelijk maakt in theorie om met een button die buiten de form staat, toch een specifiek gekozen form submitten. (Zie [hier](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/button#form))

Ik ben er niet van overtuigd dat dit een use-case is die courant is, laat staan dat we die moeten ondersteunen maar met conversie van `is="vl-button"` naar `<vl-button>` is het wel een mogelijkheid die niet meer (of moeilijk) mogelijk is.

Zouden kunnen overwegen dit attribuut dus ook hier toe te voegen en intern door te geven aan de native `<button>`.

Misschien wachten we specifieke use-cases af voor we `vl-button` uitbreiden met `form`-edge cases en is standaard form gebruik momenteel voldoende.

ksvlaanderen replied:
Voorlopig houden zoals het is, als er een use-case van de afnemers komt voegen we het dan toe.

Goldflow commented on libs/components/src/next/button/vl-button.component.ts:
```diff
@@ -0,0 +1,130 @@
+        if (this.type === 'reset') {
+            this.closest('form')?.reset();
+        }
+
```
Gelijkaardig aan bovenstaande opmerking, verliest de afnemer ook onder meer de mogelijkheid om `formnovalidate` als attribuut mee te geven (die dan zou door gegeven wordt aan `<vl-button>` intern.

Dit zou het toelaten om een form te submitten zonder de te valideren, maar dit lijkt wederom een edge-case.

Er zijn daarnaast ook nog een paar andere attributen (onder meer ivm Popover API) die niet meer aangepast kunnen worden via de API van de `<vl-button>` maar die lijken ook niet per sé zo belangrijk zolang we zelf niets met de Popover API doen.

ksvlaanderen replied:
Voorlopig houden zoals het is, als er een use-case van de afnemers komt voegen we het dan toe.

Goldflow commented on libs/components/src/next/button/stories/vl-button.stories.ts:
```diff
@@ -0,0 +1,140 @@
+    narrow: true,
+};
+
+export const ButtonLoading = ButtonTemplate.bind({});
```
Ik vermoed dat dit ook al zo werkte in de `is="vl-button"` maar naar styling is het een beetje raar dat `loading`-style eruit ziet als `disabled` styling (behalve de tekstkleur) maar toch niet `disabled` is, op functioneel vlak gezien die nog steeds events gaat uitsturen op `click`.

Eventueel dit nog documenteren bij de documentatie voor Loading variant.

ksvlaanderen replied:
Komt in orde.

Goldflow commented on libs/components/src/next/button/vl-button.css.ts:
```diff
@@ -0,0 +1,234 @@
+            }
+
+            &:hover,
+            &:active {
```
Ondanks eerdere discussies waarbij we afgesproken hebben (tegen jou mening in) om borderless styling van primary styling te behouden, vind ik het nu ik het zie in storybook, toch een beetje storend dat bij toggelen en hoveren over de button het verspringt en neig ik nu toch eerder naar jouw voorstel om ervoor te zorgen dat dit visueel niet vespringt.

Aan jou om dit eventueel nog mee te nemen of niet.
@kspeltix Wat denk jij?

![toggle-verspringen](https://github.com/milieuinfo/uigov-web-components/assets/15731348/1b3228d3-a4cc-43e0-9485-52997ed6470a)

ksvlaanderen replied:
Done

Goldflow reviewed (commented):
Good job! Code zelf ziet er goed uit.

Enkele opmerkingen toegevoegd ivm enkele mogelijke documentatie uitbreidingen en zaken die native mogelijk waren maar nu niet meer (maar waarbij het twijfelachtig is of die belangrijk zijn).

Daarnaast kom ik ook terug op eerder voorstel van jou om border toe te voegen bij Primary styles zodat het niet verspringt bij status change.

ksvlaanderen committed:
feat: UIG-2909 - vl-button-next - introductie component

vl-button-next toegevoegd
Storybook en Cypress
Form demo voorbeeld aangepast
Form data voorbeeld aangepast
Form custom validation voorbeeld aangepast
Playground-lit form voorbeeld aangepast

kspeltix commented:
:tada: This PR is included in version 1.33.0 :tada:

The release is available on [GitHub release](https://github.com/milieuinfo/uigov-web-components/releases/tag/v1.33.0)

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:

kspeltix commented:
:tada: This PR is included in version 1.33.0 :tada:

The release is available on [GitHub release](https://github.com/milieuinfo/uigov-web-components/releases/tag/v1.33.0)

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:
